core/drivers/ns16550.c

```diff
@@ -46,30 +45,27 @@
 /* uart status register bits */
 #define UART_LSR_THRE	0x20 /* Transmit-hold-register empty */
 
-static vaddr_t chip_to_base(struct serial_chip *chip)
+static void ns16550_flush(struct serial_chip *chip)
 {
 	struct ns16550_data *pd =
 		container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = io_pa_or_va(&pd->base);
 
-	return io_pa_or_va(&pd->base);
-}
-
-static void ns16550_flush(struct serial_chip *chip)
-{
-	vaddr_t base = chip_to_base(chip);
-
-	while ((io_read8(base + UART_LSR) & UART_LSR_THRE) == 0)
+	while ((serial_in(base + (UART_LSR << pd->reg_shift), pd->io_width) &
+		UART_LSR_THRE) == 0)
 		;
 }
 
 static void ns16550_putc(struct serial_chip *chip, int ch)
 {
-	vaddr_t base = chip_to_base(chip);
+	struct ns16550_data *pd =
+		container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = io_pa_or_va(&pd->base);
```

Comment:
Minor: is this really better than `chip_to_base()`? That's an unrelated change anyways, better as a different commit.

Reply:
`chip_to_base()` only allowed us to retrieve `base` value but now we need to retrieve `reg_shift` and `io_width` values as well. So in order to avoid duplicate code, I removed `chip_to_base()`.

Reply:
Ah yes I see. That's fine.